api/src/utils/async-handler.ts

```diff
@@ -0,0 +1,35 @@
+import { NextFunction, Request, Response } from 'express';
+
+export type AsyncHandler<T> = (req: Request, res: Response, ...others: any) => Promise<T>;
+export type AsyncHandlerNext<T> = (req: Request, res: Response, next: NextFunction, ...others: any) => Promise<T>;
+export type AsyncHandlerError<T> = (
+	err: Error,
+	req: Request,
+	res: Response,
+	next: NextFunction,
+	...others: any
+) => Promise<T>;
+export type AsyncMiddleware<T> = AsyncHandler<T> | AsyncHandlerNext<T> | AsyncHandlerError<T>;
+
+/**
+ * Handles promises in routes.
+ */
+function asyncHandler<T>(handler: AsyncMiddleware<T>): any {
+	if (handler.length == 2) {
+		return function (req: Request, res: Response, next: NextFunction, ...others: any) {
+			return Promise.resolve((handler as AsyncHandler<T>)(req, res, ...others)).catch(next);
+		};
+	} else if (handler.length == 3) {
+		return function (req: Request, res: Response, next: NextFunction, ...others: any) {
+			return Promise.resolve((handler as AsyncHandlerNext<T>)(req, res, next, ...others)).catch(next);
+		};
+	} else if (handler.length == 4) {
+		return function (err: Error, req: Request, res: Response, next: NextFunction, ...others: any) {
+			return Promise.resolve((handler as AsyncHandlerError<T>)(err, req, res, next, ...others)).catch(next);
+		};
+	} else {
+		throw new Error(`Failed to asyncHandle() function "${handler.name}"`);
+	}
+}
```

Comment:
instead of checking `handler.length` to return a different type of function, you can use function overloading and checking the instance of arg[0]  - it would clean up your code and make things a bit easier to read

it may look something like this below:


```ts
import http from 'http'

function asyncHandler<T>(handler: AsyncHandler<T>): AsyncHandler<T>
function asyncHandler<T>(handler: AsyncHandlerNext<T>): AsyncHandlerNext<T> 
function asyncHandler<T>(handler: AsyncHandlerError<T ): AsyncHandlerError<T>
function asyncHandler<T>(handler: AsyncMiddleware<T>): any {
       if(handler.length > 4) {
          throw new Error(`Failed to asyncHandle() function "${handler.name}"`)
       } 

       return (...args) => {
            const isReq = args[0] instanceof http.IncomingMessage
            let next =  args[2]
            
            if(!isReq){  //isError
               next = args[3]
            }           

            return Promise.resolve(handler(...args)).catch(next)
       }
}
```

fwiw your types may need to change - just copied the types you had and gave a crude example

reference of overloading: https://www.typescriptlang.org/docs/handbook/functions.html#overloads


something to think about!

Reply:
this is exactly what I'm trying to avoid, because spread operators makes function.length become 0, and this will break integration with sentry middleware hooks.